src/core/processing/qgsprocessingparameterregistry.h

```diff
@@ -0,0 +1,168 @@
+/***************************************************************************
+                         qgsprocessingparameterregistry.h
+                         -------------------------------
+    begin                : December 2016
+    copyright            : (C) 2016 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSPROCESSINGPARAMETERREGISTRY_H
+#define QGSPROCESSINGPARAMETERREGISTRY_H
+
+#include <QString>
+#include <QMap>
+
+class QWidget;
+class QgsProcessingParameter;
+
+/**
+ * \class QgsProcessingParameterAbstractMetadata
+ * \ingroup core
+ * Stores metadata about one processing parameter class.
+ * \note In C++ you can use QgsProcessingParameterMetadata convenience class.
+ * \note added in QGIS 3.0
+ */
+class CORE_EXPORT QgsProcessingParameterAbstractMetadata
+{
+  public:
+
+    /**
+     * Constructor for parameter abstract metadata. The type parameter must be unique.
+     */
+    QgsProcessingParameterAbstractMetadata( const QString& type )
+        : mType( type )
+    {}
+
+    virtual ~QgsProcessingParameterAbstractMetadata() {}
+
+    /**
+     * Returns the unique type string which identifies the parameter type.
+     */
+    QString type() const { return mType; }
+
+    /**
+     * Creates a new parameter of this type from an encoded script code line. The name, description and optional flag will have already
+     * been parsed when this function is called, so only any extra handling of the remaining definition string needs to be done by the class.
+     */
+    virtual QgsProcessingParameter* createParameterFromScriptCode( const QString& name, const QString& description, bool isOptional, const QString& definition )
+    {
+      Q_UNUSED( definition );
+      Q_UNUSED( description );
+      Q_UNUSED( name );
+      Q_UNUSED( isOptional );
+      return nullptr;
+    }
+
+    virtual QWidget* createWidget() { return nullptr; }
+
+  protected:
+    QString mType;
+};
+
+//! Function to create a widget for the parameter
+typedef QWidget*( *QgsProcessingParameterWidgetFunc )();
```

Comment:
I think it's better to base this on `std::function<QWidget*()>` ([source](http://stackoverflow.com/a/25848890/2319028))